.github/workflows/deploy.yaml

```diff
@@ -1,30 +1,161 @@
+    runs-on: ubuntu-latest
+    outputs:
+      environment: ${{ steps.set_env.outputs.environment }}
+      charm_changed: ${{ steps.check.outputs.charm_changed }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Determine environment
+        id: set_env
+        run: |
+          if [[ -n "${{ github.event.inputs.environment }}" ]]; then
+            echo "environment=${{ github.event.inputs.environment }}" >> "$GITHUB_OUTPUT"
+          elif [[ "${GITHUB_REF}" == "refs/heads/main" ]]; then
+            echo "environment=Production" >> "$GITHUB_OUTPUT"
+          else
+            echo "environment=Staging" >> "$GITHUB_OUTPUT"
+          fi
+
+      - name: Check for charm/ changes
+        id: check
+        run: |
+          if git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep '^charm/'; then
+            echo "charm_changed=true" >> "$GITHUB_OUTPUT"
+          else
+            echo "charm_changed=false" >> "$GITHUB_OUTPUT"
+          fi
+
+  pack-rock:
+    name: Pack Rock
+    runs-on: ubuntu-latest
+    needs: setup
+    environment: ${{ needs.setup.outputs.environment }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Setup LXD
+        uses: canonical/setup-lxd@main
+
+      - name: Setup rockcraft
+        run: sudo snap install rockcraft --classic
+
+      - name: Pack rock
+        run: rockcraft pack
+
+      - name: Upload rock
+        id: upload
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ env.CHARM_NAME }}-rock
+          path: ./*.rock
+
+  publish-image:
+    name: Publish OCI Image
+    runs-on: ubuntu-latest
+    needs: [setup, pack-rock]
+    environment: ${{ needs.setup.outputs.environment }}
+    outputs:
+      image_url: ${{ steps.set_image_url.outputs.image_url }}
+    steps:
+      - name: Get Rock
+        uses: actions/download-artifact@v4
+        with:
+          name: ${{ env.CHARM_NAME }}-rock
+
+      - name: Set image URL
+        id: set_image_url
+        run: echo "image_url=ghcr.io/${GITHUB_REPOSITORY}:$(date +%s)-${GITHUB_SHA:0:7}" >> $GITHUB_OUTPUT
 
-  deploy-prod:
-    if: ${{ ( github.ref == 'refs/heads/main' && github.event.workflow_run.conclusion == 'success') || github.event_name == 'workflow_dispatch' }}
+      - name: Push to GHCR
+        run: skopeo --insecure-policy copy oci-archive:$(ls *.rock) docker://${{ steps.set_image_url.outputs.image_url }} --dest-creds "canonical:${{ secrets.GITHUB_TOKEN }}"
+
+      - name: Pull image into registry
+        run: docker pull ${{ steps.set_image_url.outputs.image_url }}
+
+      - name: Upload flask app OCI image
+        env:
+          CHARMCRAFT_AUTH: ${{ secrets.CHARMHUB_TOKEN }}
+        run: |
+          sudo snap install charmcraft --classic
+          charmcraft upload-resource ${{ env.CHARM_NAME }} flask-app-image --image=$(docker images --format "{{.ID}}" | head -n 1) --verbosity=trace
```

Comment:
This approach limits the CI job to public websites as Charmhub has not support for private Charms.

An alternative approach would be to reference the `.charm` file during the `juju refresh` command which will allow for private repos to use this CI job.

Example: https://github.com/canonical/specs-v2.canonical.com/blob/main/.github/workflows/deploy.yaml#L227